File: Activity/Week4/Python/maximalSquare.py
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        if not matrix or not matrix[0]:
            return 0
            
        row, col = len(matrix), len(matrix[0])
        result = 0
        side = [0] * col
        
        for i in range(row):
            newSide = [int(matrix[i][0])] + [0 for _ in range(col - 1)]
            
            for j in range(1, col):
                if matrix[i][j] == '1':
                    newSide[j] = 1 + min(newSide[j - 1], side[j], side[j - 1])
                    
            result = max(result, max(newSide))
            side = newSide
            
        return result**2

File: Explore/2020April/Week4/C/maximalSquare.c
int min2(int a, int b)
{
    if(a < b)
        return a;
    
    return b;
}

int min3(int a, int b, int c)
{
    return min2(min2(a, b), c);
}

int max2(int a, int b)
{
    if(a < b)
        return b;
    
    return a;
}

int maximalSquare(char** matrix, int matrixSize, int* matrixColSize){
    if(matrixSize == 0)
        return 0;
    
    int **cache = malloc(sizeof(int*) * (matrixSize + 1));
    
    for(int i = 0; i <= matrixSize; i++)
        cache[i] = calloc(matrixColSize[0] + 1, sizeof(int));
    
    int result = 0;
    
    for(int row = 0; row < matrixSize; row++)
    {
        for(int col = 0; col < matrixColSize[0]; col++)
        {
            if(matrix[row][col] == '1')
                cache[row + 1][col + 1] = min3(cache[row][col + 1], cache[row + 1][col], cache[row][col]) + 1;
            
            result = max2(result, cache[row + 1][col + 1]);
        }
    }
    
    for(int i = 0; i < matrixSize; i++)
        free(cache[i]);
    
    free(cache);
    
    return result * result;
}
